### ISSUE TITLE
Connecting to Let's Encrypt: "unable to get local issuer certificate"

### ISSUE THREAD

**Issue by TBhimdi** [nzbget/nzbget#403]
I'm currently running NZBGet v19 and NZB Hydra using Let's Encrypt SSL certificates. Both work fine (green lock icon) via the web browser and worked fine together in v18 (and v19 with CertCheck=no). The issue arises when not when NZB Hydra sends NZBGet a URL (it shows up in the Downloads screen), but right after when NZBGet (I assume) is trying to connect back to NZB Hydra.

I assumed it was an NZB Hydra issue, but the author sent me here:
https://github.com/theotherp/nzbhydra/issues/623

Detailed connection and output information in the above GitHub issue.

**Comment by hugbug** [comment]
"foo.example3.com" is an indexer?

In NZBGet web-interface click on Add-button then put the URL "https://foo.example3.com" into "Add from URL" edit-box and click on **Submit**. It doesn't matter if the URL points to an nzb-file or to an html-page (such as index page of the site). NZBGet will try to fetch the remote resource and what we want to know if it succeeds at this stage or reports errors. If the fetching succeeds but URL doesn't point to an nzb-file a history item with status "SKIPPED" will be created in NZBGet.

If it fails - then try adding url "https://nzbget.net", which also has a certificate from Let's Encrypt and which works for me without errors.

**Comment by TBhimdi** [comment]
No indexers, those domains are just pointing to my VPS which run NZBGet and NZB Hydra.

I tried what you suggested.

https://foo.example.com:5075 (NZB Hydra): ERROR - "unable to get local issuer certificate"
https://foo.example.com:443 (Apache): WARNING- "HTTP/1.1 403 Forbidden" (expected, port blocked, no site)
https://nzbget.net: INFO - worked

**Comment by hugbug** [comment]
In nzbhydra issue tracker you are writing that `openssl s_client -connect foo.example.com:5075` doesn't work either.

When you connect from browser (which works without issues) do you run the browser on the same computer with NZBGet and openssl? What browser is it? If it's not Firefox, can you please try with Firefox? 

NZBGet uses the root certificates store downloaded from Mozilla web-site.

**Comment by TBhimdi** [comment]
NZBGet and NZBHydra are on a VPS, I ran that openssl command via a ssh window. When I connect from my home machine using Chrome, both 443 and 5075 work perfectly fine, no issues, green locks.

I just tried via Firefox, 443 worked, but 5075 gave an insecure site page. Both connecting to the same VPS machine, same cert, but different daemons (443 being Apache, while 5075 being NZB Hydra).

**Comment by hugbug** [comment]
So it seems the root certificate of your certificate is missing from Mozilla certificate store.

You can add that certificate into **CertStore**-file of NZBGet or disable option **CertCheck** in NZBGet as a workaround.

**Comment by TBhimdi** [comment]
Seems strange since the same cert works via Apache on 443, plus Mozilla should work fine with Let's Encrypt:
https://letsencrypt.org/2016/08/05/le-root-to-be-trusted-by-mozilla.html

I did think about adding the cert into the CertStore, but I would have to do that manually every time there's an NZBGet update. I would think NZB Hydra having the option to add a chain cert would remedy this. Until then I'll disable CertCheck, which I really did not want to do.

**Comment by hugbug** [comment]
Do you use the exactly same certificate in apache and hydra? If that's the case I would say hydra is doing something wrong.

**Comment by TBhimdi** [comment]
I noticed in Paths for CertStore it says:
"The option can point either to one big file containing all root certificates or to a directory containing certificate files, in PEM format."

So I created a subdir, sslcert, and set the CertStore path to: "${AppDir}/sslcert/"

But when I go to test a news server connection, I get this:

```
Could not set certificate store location
```

Tried it without the trialing slash as well. Am I missing how to set the correct dir? I wanted to try the Let's Encrypt PEM file from here:
https://letsencrypt.org/certificates/

**Comment by TBhimdi** [comment]
Yes, same exact certificate in Apache and Hydra (it's symlinked). And all the data points to NZB Hydra, will reach out in that GitHub issue.

If you have an inkling on what the issue could be, please post there.

**Comment by hugbug** [comment]
The syntax for CertStore is correct. It fails probably because it can't find certificates there or maybe it contwins files in wrong format. Check if you have path /etc/ssl/certs and try settings CertCheck to that path.

**Comment by TBhimdi** [comment]
I do have an "/etc/ssl/certs", but it didn't work.

"/etc/ssl/certs/" - Could not set certificate store location
"/${AppDir}/sslcert/" (with only Let's Encrypt pem) - Could not set certificate store location
"/${AppDir}/sslcert/" (with only cacert.pem) - Could not set certificate store location
"/${AppDir}/sslcert/" (with cacert.pem & Let's Encrypt pem) - Could not set certificate store location

**Comment by hugbug** [comment]
Maybe without trailing slash?

**Comment by TBhimdi** [comment]
Tried all four again without the trailing slash, same errors. CertCheck=yes for all. With CertCheck=no, they were all successful upon server test connection.

**Comment by hugbug** [comment]
Is "Could not set certificate store location" all you get in error message?

**Comment by TBhimdi** [comment]
Yes, it's the only thing in the pop-up.

**Comment by hugbug** [comment]
Setting to directory doesn't work for me either.
Looking at the code I see I messed up the init call and for now it works only with files but not with directories. Sorry for that, this will be fixed. It's a pity no one has tried this in four months it was in testing stage.

---

Can you try this same certificate in nzbget web-server and then check if Firefox and openssl can connect?

**Comment by TBhimdi** [comment]
Thanks for the future fix, appreciate it.

NZBGet uses the same cert as NZB Hydra (Let's Encrypt, set up for multiple domains, all symlinked). I've never had issues connecting to NZBGet via Chrome. Just tried with Firefox and I get a "Your connection is not secure" page.

Tried the openssl client too, same thing as connecting to NZB Hydra (works with Apache/443 still):

```
openssl s_client -connect nzbget.example.com:6791
CONNECTED(00000003)
depth=0 CN = foo.example.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = foo.example.com
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/CN=foo.example.com
   i:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
---
Server certificate
-----BEGIN CERTIFICATE-----
.
.
```

**Comment by hugbug** [comment]
Oh, great. Now we have to fix this in nzbget and then tell the hydra dev what to do, or vice versa ;)

I have an idea. Sometimes programs don't like symlinks. That's rare but who knows. Can you try with real files?

**Comment by TBhimdi** [comment]
Ha, I'm actually sorry. Two bugs were not my intention, I was hoping it was just user error on my part.

I just tried the ssl certs without the symlinks, no change.

Here's the SSL check for Let's Encrypt cert on NZBGet/6791:
https://www.sslshopper.com/ssl-checker.html#hostname=nzbget.example.com:6791
"The certificate is not trusted in all web browsers. You may need to install an Intermediate/chain certificate to link it to a trusted root certificate."

Same check using a Let's Encrypt cert from nzbget.net/443:
https://www.sslshopper.com/ssl-checker.html#hostname=nzbget.net

**Comment by hugbug** [comment]
Unfortunately I can't see what you see on sslshopper check. Can you describe the difference or provide screenshots?

Is there a way for me to get a let's encrypt certificate which I could test in nzbget on my machine (without having a publically available web-server)?

**Comment by TBhimdi** [comment]
NZBGet on port 6791:
![6791a](https://user-images.githubusercontent.com/5523497/27606222-6e76eea4-5b4d-11e7-8678-084543139cad.png)

Here's the "Learn more" link:
https://www.sslshopper.com/ssl-certificate-not-trusted-error.html

nzbget.net on port 443:
![443](https://user-images.githubusercontent.com/5523497/27606221-6e6ee830-5b4d-11e7-956c-9742ba5e0e4d.PNG)

**Comment by TBhimdi** [comment]
As for a Let's Encrypt cert for testing without a public web server, I honestly don't know, I doubt it, mainly because Let's Encrypt is pretty strict about requiring a web server on port 80 & 443 ONLY, just to issue a cert. Making one just for a test would require making a full fledged server.

For some reason, I still think it's an SSL cert chain issue that needs to be specified and used within NZBGet/Hydra alongside the other two SSL files, but that's just a guess.

**Comment by hugbug** [comment]
Yes, it looks like the chain issue although the chain support was added a long time ago in #110.

Maybe you could create a certificate for a (trash-) subdomain of your site, the certificate which you could give me for testing? Or I could create a DNS-record of some subdomain on nzbget.net to point to your server for the purpose of creating such certificate. Would that work?

If you think that's possible please send me a message to nzbget@gmail.com to discuss this privately.

**Comment by hugbug** [comment]
Thanks for the test Let's Encrypt certificate.

- I've installed this certificate in NZBGet's web-server.
- Firefox (on Windows) can connect to NZBGet without warnings/errors.
- Check with `openssl s_client -connect nzbget.example.com:6791` fails with "verify error:num=20:unable to get local issuer certificate".
- Added an URL to NZBGet download queue, the URL refers to NZBGet web-interface (https://nzbget.example.com:6791). Result: NZBGet reports an error, similar to the one reported by openssl (CertCheck=yes in nzbget).
- The reason for this error: the certificate is signed by an intermediate certificate from Let's Encrypt. That certificate isn't in NZBGet root cert store (option **CertStore**, by default file **cacert.pem** in nzbget directory).
- It's strange that it worked in Firefox because NZBGet uses the root cert store (cacert.pem) obtained from Mozilla web-site. So they both should either work or not.
- Added the Let's Encrypt intermediate certificate into cacert.pem used in NZBGet - now fetching in NZBGet works. It will also work for any other certificate signed by Let's Encrypt.
- Restored original cacert.pem (removed Let's Encrypt).
- Opened the test certificate (which we use in NZBGet web-server) in a text editor and at the end of file added the content of Let's Encrypt intermediate certificate. Now fetching in NZBGet works and we can use the default cacert.pem coming with NZBGet. `openssl s_client -connect nzbget.example.com:6791` works too.

Whether the combined certificate will work in hydra depends on how hydra loads the certificate. NZBGet uses openssl function `SSL_CTX_use_certificate_chain_file`. Older NZBGet version were using `SSL_CTX_use_certificate_file` and combined certificates did not work.

**Comment by TBhimdi** [comment]
RE: Firefox
I'm still surprised your Firefox install doesn't give the error (even though Chrome and Edge work fine). I just tried it on 3rd machine (clean VM with Firefox 54.0 portable), same certificate error page.

Have you tried Firefox on another machine? Apparently Firefox doesn't clean the certificate cache (even in Private mode):
http://community.nethserver.org/t/letsencrypt-certificate-not-secure-in-firefox-only/6108

RE: Concatenate Certs
I don't know how sustainable adding the intermediate certificate is to the end of the current certificate, mainly because it will have to be done every month or so. Let's Encrypt certs expire and renew every 30 or so days, thus overwriting the previous cert.

RE: OpenSSL
Instead of concatenating, if a separate intermediate chain file is specified (ssl.ca I think, otherwise the DST Root CA from Let's Encrypt's website), can it be added to the chain via "SSL_CTX_add_extra_chain_cert"? According to the docs, it's currently falling back to the Mozilla CA store, which doesn't contain it for whatever reason. This way the user can specify whatever intermediate cert they'd want (Let's Encrypt can't be the only cert having this problem):
https://www.openssl.org/docs/man1.0.2/ssl/SSL_CTX_add_extra_chain_cert.html

**Comment by hugbug** [comment]
**Found another interesting thing.**

Currently NZBGet uses root cert store (option **CertStore**) only in client mode, to verify certificates sent by servers to which NZBGet connects.

However, if I init the internal root cert store before loading server certificate into NZBGet's web-server, openssl finds necessary intermediate certificates in cert store and adds them into certificate chain. The full chain is then sent to clients. The results are as if you were adding all chain certificates into SecureCert-file. 

That's actually the right thing to do and I'll add this code into nzbget.

---

Why do you resist putting the intermediate certificate into NZBGet's CertStore-file? You can make a copy of this file under another name to avoid overwriting by NZBGet updates. I believe the root certificates rarely change to not care about.

I'll test how "SSL_CTX_add_extra_chain_cert" works anyway.

---

**Back to nzbhydra.**

Since it's written in python I assume the developer doesn't do any direct interaction with openssl and uses python's functions instead. That functions most likely load the full chain, not just the certificate. Python uses system openssl as encryption backend. Since your checks with `openssl s_client` fail that confirms that intermediate Let's Encrypt certificate is missing from the root cert store used by openssl (and by python). If you add it there the hydra might start using it.

Note: although NZBGet uses openssl it doesn't use system cert store, which local `openssl`-command uses and which (I assume) also uses python.

**Comment by TBhimdi** [comment]
That internal root store option sounds perfect, strange undocumented workaround, but whatever, sounds like it works.

As for the concatenation, you're correct, the CA doesn't get updated that often, I don't know why I assumed that. I just want to avoid it mainly because I don't want to add the CA to the NZBGet certstore every time NZBGet is updated, since it would overwrite the old one. Plus if I want to update it to the latest cacert.pem myself, same problem. Looks like Mozilla updates it every couple months, give or take. It's mainly a hassle I wanted to avoid.

If the "SSL_CTX_add_extra_chain_cert" option works, then maybe nzbhydra can implement that, fingers crossed.

**Comment by hugbug** [comment]
Actually although the description of option **CertCheck** in NZBGet says that it can be either file or a directory, the root cert store in openssl can use both simultaneously - the file (with a bunch of certificates) and a directory (for additional certificates).

I have to make the simultaneous usage possible. Then you would be able to keep the default "cacert.pem" (updated with each NZBGet update) and you can put extra certificates into the additional directory.

I can either split **CertStore** into two options **CertStoreFile** and **CertStoreDir** or I can keep using **CertStore** but allow entering both a file and a directory there, separated by semicolon:
```
CertStore=${AppDir}/cacert.pem;${AppDir}/mycerts
```

**Comment by hugbug** [comment]
Tested with both cacert.pem and a directory active at the same time and it worked!

The Let's Encrypt intermediate certificate must be saved with the file name `4f06f81d.0` (hash of the certificate) in the cert store directory. Try this (put the file into default location for openssl `/etc/ssl/certs`). There is a chance no changes are needed in hydra.

**Comment by TBhimdi** [comment]
Yes! That's great news. SSL is much too confusing to me, glad you got it working.

Also, from talking to you and trying to figure out what could be the issue, I dug into the NZB Hydra python code and found a fix too!

https://github.com/theotherp/nzbhydra/issues/623#issuecomment-311991761

**Comment by hugbug** [comment]
`load_verify_locations` is a function used by nzbget to init cert store. Using it load only one certiticate isn't correct as this will prevent openssl from using the cert store (if it was already set during default initialization). If the new ssl context created in hydra doesn't have any cert store set, the right thing to do were either to use `load_verify_locations` with system cert store.

**Comment by TBhimdi** [comment]
True, but if NZB Hydra is loading this SSL context as a server, why would it require any other certificate than the one specified? It will either be unencrypted or encrypted with the specified SSL cert.

I *think* NZB Hydra creates a separate context to access SSL servers as a client via the requests module, I'm not 100% on that one. Will let the NZB Hydra dev chime in to correct me.

Either way, I'll see if a ```load_default_certs``` after ```load_verify_locations``` works. Maybe it won't overwrite the manually specified intermediate cert.

**Comment by TBhimdi** [comment]
Looks like it didn't overwrite and it still works. ```load_default_certs```, after manually specifying the cafile, doesn't break anything with Let's Encrypt and insures OS level certs are still loaded. Thanks for pointing that out.

**Comment by hugbug** [comment]
Closing this issue. The TODOs will be tracked in #411.

**Comment by satmandu** [comment]
FYI this appears broken in the 20.0-testing-r2075 build whereas let's encrypt ssl worked in 19.1 release build.

I see this error from another computer, but nothing in the nzbget logs:

```
openssl s_client -connect nzbget.local:6791 -prexit -showcerts
CONNECTED(00000003)
depth=1 /C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
verify error:num=20:unable to get local issuer certificate
verify return:0
```

(Where SecureCert is pointed to the fullchain.pem and SecureKey  is pointed to the privkey.pem.)
This worked fine in 19.1, with nothing else changed.

**Comment by hugbug** [comment]
Nothing was changed in nzbget either. Except updated root store file.
Make a test in your working 19.1 installation by putting nzbget-binary from 20.0 there and not updating any other files/config.

**Comment by satmandu** [comment]
Ok I tried that. I kept the nzbget.conf the same.

SSL Working: WebDir in nzbget.conf pointed to the 19.1 webui folder. 
Using:
19.1 nzbget binary

SSL not Working: WebDir in nzbget.conf pointed to the 19.1 webui folder. 
Using:
20.0 nzbget binary

SSL Not working: WebDir in nzbget.conf pointed to the 20.0 webui folder. 
Using:
19.1 nzbget binary
20.0 nzbget binary


Interestingly, I'm getting the same openssl error with the working configuration as with the non-working configuration:
```
openssl s_client -connect nzbget.local:6791 -prexit -showcerts
CONNECTED(00000003)
depth=1 /C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
verify error:num=20:unable to get local issuer certificate
verify return:0
```

(Tested from Chrome & Safari.)
